modin/core/dataframe/algebra/binary.py

```diff
@@ -94,11 +95,28 @@ def caller(
                         )
                     )
                 else:
+                    boolean_operators = ["eq", "ge", "gt", "le", "lt", "ne"]
```

Comment:
I'm a little wary of hardcoding the list of operators in this fashion--though it's not likely to change, it does risk bugs in case there's an edge case for a binary operator that we haven't checked.

Can you add a flag to `Binary.caller` that controls this behavior instead? Something like
```
def caller(query_compiler, [etc...], dtypes=None, precompute_dtypes=False, result_dtype_is_bool=False):
```

where the [corresponding locations in the query compiler](https://github.com/modin-project/modin/blob/882433c46ea59df9850b11b4c6088467a95a1a8f/modin/core/storage_formats/pandas/query_compiler.py#L378) would then look something like this:
```
add = Binary.register(pandas.DataFrame.add, precompute_dtypes=True)
...
eq = Binary.register(pandas.DataFrame.eq, result_dtype_is_bool=True)
```

This per-operator configuration is easier to make changes for, compared to hardcoding them in a list here.

Reply:
@noloerino, do you think if it is better to precompute dtypes at the query compiler layer? Then we could avoid this hardcoding, as well as branching with checks for a function name.

Reply:
@YarShev @noloerino I have changed the logic for precomputing dtypes to query compiler as suggested by @YarShev . Currently I have added logic only for add function. 
Would this be a better option considering the code would be repeated(maybe as a function call) once for each binary function(~30 times)?

Reply:
I think the actual code that does the precomputing should be in the operator `caller` method to avoid repetition, but configured in the query compiler with flags in the manner I suggested. That would avoid both the repetition and the hardcoding.

Reply:
I still don't feel that the precomputing should be in the operator `caller` method. The flags `precompute_dtypes` and `result_dtype_is_bool` seem to me contradicting a bit (Say when `precompute_dtypes is False` and `result_dtype_is_bool is True`. In addition, we still would have the branching in depend on a concrete binary function (check for `truediv` is already presented in this PR).

> Would this be a better option considering the code would be repeated(maybe as a function call) once for each binary function(~30 times)?

I think a function call once for each binary function doesn't seem horrible.

Considering that we have already been passing dtypes from the qc layer in to map operator, for instance, as well as this text from `Type Rules by Operator` section of https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf,

```
For operators that allow a UDF/SDF as input, the output types can
be specified by the developer (i.e., “Specified” in Table 1), or left
unspecified. Suppose a developer wants to perform a map with a
SDF that returns TRUE for non-NULL values, and FALSE otherwise.
Since all columns in the output are known to be boolean, this information can be provided by the developer implementing the SDF
to the map upfront to avoid costly type inference. Alternatively,
when the types of the output dataframe after a map is not known, it
ends up being UNSPECIFIED for every column.
```

I think the precomputing dtypes should  happend at the qc.

cc @modin-project/modin-core, other thoughts?

Reply:
@noloerino suggestion implemented in this PR, my suggestion in #5525.

@modin-project/modin-core, do you think which approach we should go with?

Reply:
> text from Type Rules by Operator section

Good reference, although I think it's more related to actual dataframe algebra defined in a low-level dataframe object. Our "algebra" module is more like wrappers around actual operators, thus I think it's acceptable here to pass instructions on how to get new dtypes instead of already computed ones.

Regarding where to put the inferring logic: I believe it would be better to encapsulate it in the Binary operator. Since the logic tends to be only binary ops related it's then unnecessary to bloat general QC with operator-specific functionality.

As for configuring the type-inferring logic, I would propose something like this. A parameter for `.register` that either takes a resulted type (bool for logical ops, float for `truediv`, ...) or an "infer" keyword, indicating automatic type inferring.
```python
dtypes : {int, float, bool, object, 'infer', None}, default: None
    Specifies a dtype for the result of the binary operation. If 'infer' - try to
    automatically infer the resulted type based on the operand's types. If ``None``
    - do not infer new dtypes (they will be computed manually once accessed).
```
If you don't like mixing actual types and 'infer' in a single parameter we could either reformat the parameter a bit:
```python
how_compute_dtypes : enum.HOW_INFER_TYPE = {INT, FLOAT, BOOL, NOHOW, AUTO}
```
or split them into two separate parameters:
```python
def register(..., dtypes=None, auto_dtypes_inferring=False): # 'dtypes' is ignored if 'auto_dtypes_inferring=True' was specified
```

Reply:
In order not to bloat QC module we wanted to have kernels.py file where we would define respective kernels and would just import those into QC to keep it clean (I do not remember the exact issue for this). We already have the QC bloated enough with some kernels. Also, we may want to compute dtypes for other operators as well and it might be complicated to do that at the algebra layer as here. Each operator is not aware of a function to be executed by it so that might be complex to infer dtypes at this layer. In addition, I am afraid of possible multiple branching in depend on a concrete function in every operator. I still feel we should compute dtypes at the QC layer. @dchigarev, thoughts/comments?

I also want to listen to thoughts from someone from ponder team. @mvashishtha, @RehanSD, @vnlitvinov, @devin-petersohn

Reply:
The way I typically think about dtype handling is that it should be controlled by the algebra layer (what we called the Modin Dataframe layer previously), but the QC layer will need to give hints or suggestions.

We have logical operators and physical operators. The algebra are logical, they call the physical operators from the partition manager.

Reply:
The query compiler is supposed to have knowledge of the compute kernels and in-memory format of the data. It gives hints to the underlying layer (Modin Dataframe layer). We have the structure isolated from each other but the code may become too complicated (for instance, like here the code of the algebra layer may become complicated). And so we will have to expand the knowlodge of a concrete layer in future if we want the code to be clean. These are my thoughts regarding this. So, let's move forward with the suggestion above and see how it goes. We can always refactor the code. @arunjose696, please push this forward with @dchigarev suggestion. I think it looks clearer than adding `precompute_dtypes` and `result_dtype_is_bool` flags.

> As for configuring the type-inferring logic, I would propose something like this. A parameter for .register that either takes a resulted type (bool for logical ops, float for truediv, ...) or an "infer" keyword, indicating automatic type inferring.

```python
dtypes : {int, float, bool, object, 'infer', None}, default: None
    Specifies a dtype for the result of the binary operation. If 'infer' - try to
    automatically infer the resulted type based on the operand's types. If ``None``
    - do not infer new dtypes (they will be computed manually once accessed).
```

Reply:
I would say that dtypes _inference_ should probably be handled by QC (inner layer), but dtypes _knowledge_ seems to be tied more to outer layer, i.e. closer to the real data structures (QC should be more abstract).

Hence I think we should define dtypes at the algebra layer.

As for the interface, I propose to introduce two `object()` instances and use them as values so as to not mix strings and `None`-s, like
```python
InferDtypes = object()
ComputeOnAccess = object()
# skip
def register(..., dtypes=...):
    # skip
    if dtypes is InferDtypes:
        # infer them
    elif dtypes is ComputeOnAccess:
        # do not infer
    else:
        # dtypes were specified explicitly
```

This works much like `pandas._libs.lib.NoDefault` and is super-explicit.

Reply:
I prefer to go with a enumerator parameter how_compute_dtypes as in @dchigarev  suggestion as mixing actual types and 'infer'  might be confusing.

> ```
> 
> If you don't like mixing actual types and 'infer' in a single parameter we could either reformat the parameter a bit:
> 
> ```python
> how_compute_dtypes : enum.HOW_INFER_TYPE = {INT, FLOAT, BOOL, NOHOW, AUTO}
> ```
>

Reply:
I am okay with @dchigarev suggestion I pointed, as well as @vnlitvinov suggestion but I wouldn't introduce `ComputeOnAccess = object()` for now because we use `None` for dtypes everywhere to compute it on access.

Reply:
I like the suggestion @arunjose696 and @YarShev provide - the algebra layer is agnostic to what specific functions are being called (e.g. if we have a map whose function is to convert the dtype of all columns to float), while the query compiler is often the one specifying what these kernels are (in this case the function that converts all values to float), so it makes sense for the QC to tell the algebra layer hints about the dtype (either the specific types that will result, or to infer the type (if it is necessary, e.g. during query composition) or that we don't know what the resultant types will be, and we don't need to defer them - the "NOHOW" suggestion as I understand).